### PR TITLE
[SPARK-22639][SQL] Support aggregate cbo stats estimation if the group by clause involves substring

### PR DESCRIPTION
**Pull request by shahidki31**
### What changes were proposed in this pull request?
If the grouping keys contains substring, Aggregate doesn't use CBO for computing stats and fall back to sizeInBytes based estimation. So we will be missing row count and col stats after that. Many TPCDS queries has substring in their group by expressions, like Query8, Query23a etc.

### Why are the changes needed?
CBO cannot estimate rowcount if the groupby clause of a query involves the expression substring. For example, we can not estimate the row count of the following query, 

**Eg: Query**
TPCDS Q8

**Before:**
(From TPCDS Q8 plan)
No rowcount
```

 +- Aggregate [ca_zip#0], [ca_zip#0], **Statistics(sizeInBytes=160.2 MiB)**
       +- Project [substr(ca_zip#94, 1, 5) AS ca_zip#0], Statistics(sizeInBytes=160.2 MiB, rowCount=6.00E+6)
```
**After:**

Rowcount exists

```
+- Aggregate [ca_zip#0], [ca_zip#0], **Statistics(sizeInBytes=241.6 KiB, rowCount=9.90E+3)**
       +- Project [substr(ca_zip#94, 1, 5) AS ca_zip#0], Statistics(sizeInBytes=143.1 MiB, rowCount=6.00E+6)

```
### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added UTs, manual testing

Imapct on TPCDS Query8:
SF=1000, 8 node cluster with ExecMemory 28GB, ExecCores 4, ExecInstances 15.

<meta name="ProgId" content="Excel.Sheet">
<meta name="Generator" content="Microsoft Excel 15">
<link id="Main-File" rel="Main-File" href="file:///C:/Users/shaki/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel="File-List" href="file:///C:/Users/shaki/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

Metrics | Before | After | Improvement
-- | -- | -- | --
minRunTime | 10.73   sec | 9.61 sec | 10.50%
ShuffleRead | 735 MB | 658 MB | 10.40%
Shuffle Write | 735 MB | 658 MB | 10.40%

### CODE REVIEW

**Comment by shahidki31** [comment]
cc @maropu @srowen Kindly review

**Review by srowen — COMMENTED** [review]
Is there any way to just enable this for all similar expressions rather than special case substring? Do they have a common super type and is it valid for all of them?

**Comment by shahidki31** [comment]
Sure @srowen Let me check if we can include more cases

**Comment by shahidki31** [comment]
Hi @srowen ,
  I checked the code and I ran all the TPCDS queries to find the potential cases. Following are the observations.
1. Aggregate expressions like sum, avg, count etc. -> We can't estimate stats for this.
2. Literal Expressions -> We can estimate stats. (q5 and q14b of TPCDS)
3.  String repeat -> Opposite of substring. We can estimate the stats.

I have added support for 2 and 3. I am not sure we can find a supercase, as each one's estimation will be different. 
Please review. Thanks

**Comment by srowen** [comment]
I don't know enough to fully review this. Seems plausible but seems also pretty narrowly scoped, but I don't know.

**Comment by shahidki31** [comment]
Thanks @srowen .

**Comment by shahidki31** [comment]
cc @wzhfy @cloud-fan Kindly review

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala [hunk 1]
```diff
@@ -80,6 +80,54 @@ object EstimationUtils {
...
         alias.toAttribute -> attributeStats(attr)
+      case alias @ Alias(expn: Expression, _) if isExpressionStatsExist(expn, attributeStats) =>
+        getExpressionStats(alias.toAttribute, expn, attributeStats)
+    }
+  }
+
+  // Support for substring expressions.
+  // TODO: Support for more expressions like Multiply.
+  private def isExpressionStatsExist(
+      expn: Expression,
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala:91]: `expn` -> `expr`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala [hunk 2]
```diff
@@ -80,6 +80,54 @@ object EstimationUtils {
     expressions.collect {
       case alias @ Alias(attr: Attribute, _) if attributeStats.contains(attr) =>
         alias.toAttribute -> attributeStats(attr)
+      case alias @ Alias(expn: Expression, _) if isExpressionStatsExist(expn, attributeStats) =>
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala:83]: Why did you update this method instead of `AggregateEstimation`? `Project` uses this method though. Is this related to projections?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala [hunk 3]
```diff
@@ -80,6 +80,54 @@ object EstimationUtils {
...
+        getExpressionStats(alias.toAttribute, expn, attributeStats)
+    }
+  }
+
+  // Support for substring expressions.
+  // TODO: Support for more expressions like Multiply.
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala:89]: Why do we need to handle individual exprs here? For aggregate stat estimation, we cannot just use upper-bould stat values from a child plan in `AggregateEstimation`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala (hunk 2 above)
Comment (shahidki31) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala:83]: Yes, We need to update for project node as well. If the grouping key is aliased, or if the grouping key is complex expression, the expressions are pushed to project node.
Even if the aggregate node is not there, still we need to update the stats for expressions in the project node.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala (hunk 3 above)
Comment (shahidki31) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala:89]: For substring, yes we can pass the child plan stats. But for other expressions it might not be true. For eg: Literal, Concat string etc. We might need to find more expressions if we can estimate stats. For eg: Multiply.

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala:89]: Why can we keep discount stats for substring?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala (hunk 2 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala:83]: Ah I see.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala (hunk 3 above)
Comment (shahidki31) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala:89]: Sure. I can update stats for substring. Basically max len and avg length can update. Null count remains same. Distinct count will remain upper bound. Actual distinct count can't estimate after applying substring
